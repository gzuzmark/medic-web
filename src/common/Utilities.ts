const donwloadLink = (link: string, filename: string, ext: string) => {
    const tempLink = document.createElement('a');
    tempLink.style.display = 'none';
    tempLink.href = link;
    tempLink.setAttribute('download', `${filename}.${ext}`);

    // Safari thinks _blank anchor are pop ups. We only want to set _blank
    // target if the browser does not support the HTML5 download attribute.
    // This allows you to download files in desktop safari if pop up blocking
    // is enabled.
    if (typeof tempLink.download === 'undefined') {
        tempLink.setAttribute('target', '_blank');
    }

    document.body.appendChild(tempLink);
    tempLink.click();
    document.body.removeChild(tempLink);
};

const getDateFormatted = (date: Date) => {
    const dd = date.getDate();
    const mm = date.getMonth() + 1;
    let day = dd.toString();
    let month = mm.toString();
    const year = date.getFullYear();

    if(dd < 10){
        day = '0' + dd;
    }
    if(mm < 10){
        month = '0' + mm;
    }
    return day + '-' + month + '-' + year;
}

const Utilities = {
    donwloadLink,
    getDateFormatted
};

export default Utilities;