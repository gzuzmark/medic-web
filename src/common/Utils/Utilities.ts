import * as moment from 'moment';
import defaultProfile from '../../assets/images/default.png';

const donwloadLink = (link: string, filename: string, ext: string) => {
    const tempLink = document.createElement('a');
    tempLink.style.display = 'none';
    tempLink.href = link;
    tempLink.setAttribute('download', `${filename}.${ext}`);

    // Safari thinks _blank anchor are pop ups. We only want to set _blank
    // target if the browser does not support the HTML5 download attribute.
    // This allows you to download files in desktop safari if pop up blocking
    // is enabled.
    if (typeof tempLink.download === 'undefined') {
        tempLink.setAttribute('target', '_blank');
    }

    document.body.appendChild(tempLink);
    tempLink.click();
    document.body.removeChild(tempLink);
};

const getDateFormatted = (date: Date) => {
    const dd = date.getDate();
    const mm = date.getMonth() + 1;
    let day = dd.toString();
    let month = mm.toString();
    const year = date.getFullYear();

    if(dd < 10){
        day = '0' + dd;
    }
    if(mm < 10){
        month = '0' + mm;
    }
    return day + '-' + month + '-' + year;
};

const getValue = (value?: string, backup?: string): string  => {
    if (!value && backup) {
        value = backup;
    }
    return value ? value : '';
};

const getDocumentHeight = (): number => {
    const body = document.body;
    const html = document.documentElement;
    return Math.max(
        body.scrollHeight,
        body.offsetHeight,
        html.clientHeight,
        html.scrollHeight,
        html.offsetHeight);
};

const scrollToTop = () => {
    if (!!window && !!window.scrollTo) {
        window.scrollTo(0, 0);
    }
};

const doClone = (source: any): any => {
    if (Object.prototype.toString.call(source) === '[object Array]') {
        const clone = [];
        for (let i=0; i<source.length; i++) {
            clone[i] = doClone(source[i]);
        }
        return clone;
    } else if (typeof(source) === "object") {
        const clone = {};
        for (const prop in source) {
            if (source.hasOwnProperty(prop)) {
                clone[prop] = doClone(source[prop]);
            }
        }
        return clone;
    } else {
        return source;
    }
};

const getMonday = (date?: Date) => {
    let d = Utilities.todayDate();
    if (!!date) {
        d = new Date(date);
    }
    const day = d.getDay();
    const diff = d.getDate()  - day + (day === 0 ? - 6 : 1);
    return new Date(d.setDate(diff));
};

const todayDate = () => {
    const date = new Date();
    date.setSeconds(0);
    date.setMinutes(0);
    date.setHours(0);
    date.setMilliseconds(0);
    return date;
};

const onErrorStudentImage = (e: any) => {
    e.target.onerror = null;
    e.target.src = defaultProfile;
};

const deepEqual = (a: any,b: any) => {
    if( (typeof a === 'object' && a != null) && (typeof b === 'object' && b != null) ) {
        const count = [0, 0];
        for(const key in a) {
            if (key in a) {
                count[0]++
            }
        }
        for( const key in b) {
            if (key in b) {
                count[1]++;
            }
        }
        if( count[0]-count[1] !== 0) {
            return false;
        }
        for( const key in a) {
            if(!(key in b) || !deepEqual(a[key],b[key])) {
                return false;
            }
        }
        for( const key in b) {
            if(!(key in a) || !deepEqual(b[key],a[key])) {
                return false;
            }
        }
        return true;
    } else {
        return a === b;
    }
};

const getAgeByBirthDate = (bd: string) => bd ? moment().diff(bd, 'years', false) : '';

const Utilities = {
    deepEqual,
    doClone,
    donwloadLink,
    getAgeByBirthDate,
    getDateFormatted,
    getDocumentHeight,
    getMonday,
    getValue,
    onErrorStudentImage,
    scrollToTop,
    todayDate
};

export default Utilities;
