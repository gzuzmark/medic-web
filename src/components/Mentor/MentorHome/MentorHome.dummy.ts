

export const url = 'me/sessions?';
const data = {
    "items": [
        {
            "availability": {
                "count": 0,
                "limit": 3
            },
            "credits": 0,
            "from": "2018-10-26T10:12:26.000-05:00",
            "id": "c5f0daac-6729-43a1-a077-1ea4dbd0540d",
            "isActive": false,
            "isEnabledForAttendance": true,
            "location": {
                "type": "VIRTUAL",
                "typeLabel": "Virtual"
            },
            "skill": {
                "id": "a4f80320-10df-4e40-bee0-b85930ebd7c6",
                "name": "Visual Designer II"
            },
            "status": "AVAILABLE",
            "to": "2018-10-26T10:12:27.000-05:00"
        },
        {
            "availability": {
                "count": 0,
                "limit": 30
            },
            "credits": 0,
            "from": "2018-10-27T07:30:00.000-05:00",
            "id": "cea48549-d61b-44e0-865e-3c34e7c00a16",
            "isActive": true,
            "isEnabledForAttendance": false,
            "location": {
                "location": {
                    "address": "Av. Oriental 585",
                    "id": "2",
                    "location": "Av. Oriental 585, Torres Tecnológicas, Piso 4, Aula 405",
                    "maxStudents": 30,
                    "room": "Piso 4, Aula 405",
                    "site": "Torres Tecnológicas",
                    "status": "PUBLISHED"
                },
                "type": "PHYSICAL",
                "typeLabel": "Presencial"
            },
            "skill": {
                "id": "a4f80320-10df-4e40-bee0-b85930ebd7c6",
                "name": "Visual Designer II"
            },
            "status": "AVAILABLE",
            "to": "2018-10-27T07:45:00.000-05:00"
        }
    ]
};

const emptyData = {
    "items": []
};

const Dummy = {
    data,
    emptyData,
    url
};

export default Dummy;