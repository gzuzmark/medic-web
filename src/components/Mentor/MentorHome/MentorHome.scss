.MentorHome {
  margin-top: 100px;
  &_title {
    align-items: center;
    display: flex;
    justify-content: center;
    margin-bottom: 40px;
    text-align: center;
    width: 100%;
    svg, h2 {
      display: inline-block;
    }
    svg {
      height: 24px;
      margin-right: 8px;
      width: 24px;
    }
  }
}