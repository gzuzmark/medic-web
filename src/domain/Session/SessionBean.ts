import {AbstractDateParser} from "../DateManager/AbstractDateParser";

export const SESSION_STATUS = {
    ATTENDED: 'ATTENDED',
    EXPIRED: 'CADUCED',
    NO_ATTENDED: 'NO_ATTENDED',
    RATED: 'RATED',
    SCHEDULED: 'SCHEDULED',
};

export interface ISessionItemBase {
    id: string;
    name: string;
}

export interface ISessionLocation {
    typeLabel: string;
    type: string;
    location?: {
        id: string,
        site: string,
        room: string,
        address: string,
        status: string,
        maxStudents: number,
        location: string
    }
    sessionUrl?: string;
}


export interface ISessionBase {
    id?: string;
    from: string;
    to: string;
    location?: ISessionLocation;
    skill?: ISessionItemBase;
    isActive?: boolean;
}
export const SESSION_VIRTUAL = "VIRTUAL";
export const SESSION_UNDEFINED = "UNDEFINED";
export const SESSION_PHYSICAL = "PHYSICAL";

export class SessionBean {
    public session: ISessionBase;

    constructor(session: ISessionBase) {
        this.session = session;
    }

    public getTime(dateFormatter: AbstractDateParser): string {
        return `${dateFormatter.parseDateToString(this.session.from, "h:mm a")} - ${dateFormatter.parseDateToString(this.session.to, "h:mm a")}`;
    }

    public isVirtual(): boolean {
        return !!this.session.location && this.session.location.type === SESSION_VIRTUAL;
    }

    public getLocation(): string {
        let location = '';
        if (!!this.session.location && this.isVirtual()) {
            location = this.session.location.sessionUrl ? this.session.location.sessionUrl : '';
        } else if (this.session.location) {
            location = this.session.location.location ? this.session.location.location.location : '';
        }

        return location;
    }

    public getStatus() {
        let text = 'pending';
        const from = new Date(this.session.from);
        const to = new Date(this.session.to);
        const now = new Date();
        if (now.getTime() <= to.getTime() && now.getTime() >= from.getTime()) {
            text = 'active';
        } else if (!this.session.isActive) {
            text = 'resolve';
        }
        return text;
    }
}
