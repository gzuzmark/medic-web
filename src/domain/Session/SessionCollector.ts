import { SessionBean } from "./SessionBean";

export interface IBoxDayDescription {
    topText: string;
    mainText: string;
    bottomText: string;
}

export interface ISessionCollector<T> {
    description: IBoxDayDescription;
    status: string;
    date: string;
    pending_sessions: T[];
    resolve_sessions: T[];
}

export class SessionCollector<T extends SessionBean> {
    public sessions: T[];
    public sessionCollector: Array<ISessionCollector<T>>;
    public monthFormatter = new Intl.DateTimeFormat("es", { month: "long" });
    public selectedDate: string;
    public firstEnableDate: string;
    constructor(sessions: T[], selectedDate: string) {
        // , private createSession: { new(...args : any[]): T; }
        // new this.createSession(sessions);
        this.sessions = sessions;
        this.selectedDate = selectedDate;
        this.sessionCollector = [];
        this.firstEnableDate = "";
        this.buildStructure();
        this.filterSessions(this.sessions);
    }

    public getSessionsFrom(day: number) {
        return this.sessionCollector[day];
    }

    public getRangeDays() {
        return this.sessionCollector.map((item: ISessionCollector<T>) => {
            return {
                date: item.date,
                description: item.description,
                status: item.status
            }
        })
    }

    public getFirstDate(currentWeek: boolean) {
        return currentWeek ?
            new Date() :
            new Date(this.firstEnableDate);
    }

    private filterSessions(sessions: T[]) {
        sessions.forEach((item: T) => {
            const date = new Date(item.session.from);
            const collector = this.sessionCollector[date.getDay()];
            collector.status = "default" ;
            if (item.session.isActive) {
                collector.pending_sessions.push(item);
            } else {
                collector.resolve_sessions.push(item);
            }
        });
    }

    private buildStructure() {
        const date = new Date(this.selectedDate);
        for (let index = 0; index < 7; index++) {
            if (index === 1) {
                this.firstEnableDate = date.toString();
            }
            const month = this.monthFormatter.format(date);
            this.sessionCollector[index] = {
                date: date.toISOString(),
                description: {
                    bottomText: month[0].toUpperCase() + month.slice(1),
                    mainText: date.getDate().toString(),
                    topText: this.getStringDay(index)
                },
                pending_sessions: [] as T[],
                resolve_sessions: [] as T[],
                status: "disabled"
            };
            date.setDate(date.getDate() + 1);
        }
    }

    private getStringDay(index: number): string {
        let day = '';
        switch (index) {
            case 0:
                day = 'Domingo';
                break;
            case 1:
                day = 'Lunes';
                break;
            case 2:
                day = 'Martes';
                break;
            case 3:
                day = 'Miércoles';
                break;
            case 4:
                day = 'Jueves';
                break;
            case 5:
                day = 'Viernes';
                break;
            case 6:
                day = 'Sábado';
                break;
        }
        return day;
    }
}