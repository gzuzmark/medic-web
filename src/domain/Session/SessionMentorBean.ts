import { ISessionBase, SessionBean } from "./SessionBean";

export interface ISessionAvailability {
    limit: number;
    count: number;
}

export interface ISessionMentor extends ISessionBase {
    availability?: ISessionAvailability;
    isEnabledForAttendance?: boolean;
    status?: string;
}

export class SessionMentorBean extends SessionBean {
    public session: ISessionMentor;

    constructor(session: ISessionMentor) {
        super(session);
        this.session = session;
    }

    public getAvailability(): string {
        let text = 'alumno inscrito';
        let count = 0;
        let limit = 0;
        if (this.session.availability) {
            count = this.session.availability.count;
            limit = this.session.availability.limit;
        }

        if (this.session.location && this.session.location.location && this.session.availability) {
            count = this.session.availability.count;
            limit = this.session.location.location.maxStudents;
        }

        if (limit > 1) {
            text = 'alumnos inscritos'
        }
        return `${count} de ${limit} ${text}`;
    }
}