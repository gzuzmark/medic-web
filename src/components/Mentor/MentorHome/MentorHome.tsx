import * as React from 'react';
import Icon from "../../../common/Icon/Icon";
import Layout from "../../../common/Layout/Layout";
import Loader from "../../../common/Loader/Loader";
import { Headline1 } from '../../../common/MentorText';
import Utilities from "../../../common/Utils/Utilities";
import {MomentDateParser} from "../../../domain/DateManager/MomentDateParser";
import {ListenerFirebase} from "../../../domain/Listener/ListenerFirebase";
import {IBoxDayDescription, SessionCollector} from "../../../domain/Session/SessionCollector";
import {ISessionMentor, SessionMentorBean} from "../../../domain/Session/SessionMentorBean";
import UserRepository from "../../../repository/UserRepository";
import SessionService from "../../../services/Session/Session.service";
import DayHandlerBar, {IDayHandlerBar} from "./components/DayHandlerBar/DayHandlerBar";
import ModalTakeList from "./components/ModalTakeList/ModalTakeList";
import SessionsMentorDetail, {ISessionMentorDetail} from "./components/SessionsMentorDetail/SessionsMentorDetail";
import './MentorHome.scss';

export interface IRangeDay {
    status: string;
    description: IBoxDayDescription;
    date: string;
}

interface IStateMentorHomeCore {
    daysBar: IDayHandlerBar;
    loading: boolean;
    selectedDate: string;
    sessionDetail: ISessionMentorDetail;
}

interface IUseHandlerNoAttendedsessions {
    loading: boolean;
    session: SessionMentorBean | null;
    doRequest: () => void;
}

interface IPropsMentorHomeCore {
    noAttendedSessions: IUseHandlerNoAttendedsessions;
    sessionCollector: SessionCollector<SessionMentorBean>;
    setSessionCollector: (s: SessionCollector<SessionMentorBean>) => void
}

const useHandlerNoAttendedSessions = (): IUseHandlerNoAttendedsessions => {
    const [loading, setLoading] = React.useState(true);
    const [session, setSession] = React.useState(null as null | SessionMentorBean);
    const sessionService = new SessionService();
    const doRequest = () => {
        sessionService.listNoAttendedSessions().then((sessions: ISessionMentor[]) => {
            const item = new SessionMentorBean(sessions[0]);
            setSession(item);
            setLoading(false);
        }).catch(() => {
            setSession(null);
        })
    };
    React.useEffect(() => {
        doRequest();
    }, [0]);
    return {loading, session, doRequest}
};

export class MentorHomeCore extends React.Component<IPropsMentorHomeCore, IStateMentorHomeCore> {
    public state: IStateMentorHomeCore;
    private sessionService = new SessionService();
    private mentorId: string;
    private listenerFirebase: ListenerFirebase;
    private mdp: MomentDateParser;
    private interval: any = 0;
    constructor(props: IPropsMentorHomeCore) {
        super(props);
        this.state = {
            daysBar: this.updateDaysBar(0),
            loading: true,
            selectedDate: (Utilities.todayDate()).toISOString(),
            sessionDetail: {
                scrollTop: false,
                sessions: null,
            }
        };
        this.getNewSessions = this.getNewSessions.bind(this);
        this.updateDate = this.updateDate.bind(this);
        this.loadSessions = this.loadSessions.bind(this);
        this.updateScrollTop = this.updateScrollTop.bind(this);
        this.buildRefURL = this.buildRefURL.bind(this);
        this.updateSessionDetail = this.updateSessionDetail.bind(this);
        this.refreshSession = this.refreshSession.bind(this);
        this.updateDaysBar = this.updateDaysBar.bind(this);
        this.firstLoad = this.firstLoad.bind(this);
        this.mdp = new MomentDateParser();
    }

    public componentDidMount() {
        if (UserRepository.getToken() && UserRepository.getUser()) {
            this.mentorId = UserRepository.getUser().rolId;
            this.firstLoad();
        }
    }

    public shouldComponentUpdate(nextProps: IPropsMentorHomeCore, nextState: IStateMentorHomeCore) {
        return !Utilities.deepEqual(nextProps, this.props) || !Utilities.deepEqual(nextState, this.state);
    }

    public componentWillUnmount() {
        this.listenerFirebase.stopListener();
    }

    public firstLoad() {
        const date = Utilities.getMonday();
        const ref = this.buildRefURL(date);
        this.listenerFirebase = new ListenerFirebase(ref);
        this.listenerFirebase.setCallback(() => {
            const { from, to } = this.getRangeDates(date.toISOString());
            this.loadSessions(from, to, 0, true);
        });

        this.setState({
            loading: true
        }, () => {
            this.getNewSessions(date.toISOString(), 0);
            this.listenerFirebase.startListener();
        });
    }
    public render() {
        return <Layout title={"Tutores"}>
            <ModalTakeList item={this.props.noAttendedSessions.session}
                           loadNoAttendedSessions={this.props.noAttendedSessions.doRequest} />
            <div className="MentorHome u-LayoutMentorMargin">
                <div className={"MentorHome_title"}>
                    <Icon name={"calendar"}/>
                    <Headline1>Tus sesiones</Headline1>
                </div>
                <DayHandlerBar
                    onChangeDate={this.updateDate}
                    onChangeWeek={this.getNewSessions}
                    selectedDate={this.state.selectedDate}
                    loading={this.state.loading}
                    daysBar={this.state.daysBar} />
                {this.state.sessionDetail.sessions ?
                <SessionsMentorDetail
                    sessionDetail={this.state.sessionDetail}
                    selectedDate={this.state.selectedDate} /> :
                <Loader style={{marginTop: 150}}/>}
            </div>
        </Layout>
    }

    private buildRefURL(date: Date): string {
        const listenerDate = date.toISOString().replace('.', '_');
        return `mentors/${this.mentorId}/sessions/week-changes/on-start/${listenerDate}`;
    }

    private updateDate(selectedDate: string) {
        const date = new Date(selectedDate);
        this.refreshSession(date);
        this.setState({
            selectedDate,
            sessionDetail: this.updateSessionDetail(date)
        }, () => {
            const monday = Utilities.getMonday(date);
            const ref = this.buildRefURL(monday);
            this.listenerFirebase.setCallback(() => {
                const { from, to } = this.getRangeDates(monday.toISOString());
                this.loadSessions(from, to, 0, true);
            });
            this.listenerFirebase.updateRef(ref);
            this.updateScrollTop();
        })
    }

    private getNewSessions(date: string, counter: number) {
        const { from, to } = this.getRangeDates(date);
        const currentCounter = this.state.daysBar.counter + counter;
        this.setState({
            loading: true
        }, () => {
            this.loadSessions(from, to, currentCounter);
        });
    }

    private getRangeDates(date: string) {
        const from = new Date(date);
        const to = new Date(date);
        to.setDate(to.getDate() + 7);
        return { from, to }
    }

    private loadSessions(from: Date, to: Date, counter: number, forceRefresh?: boolean) {
        this.sessionService.listMentorSessions(from.toISOString(), to.toISOString())
            .then((sessions: ISessionMentor[]) => {
                let newState = {};
                const mentorSessions = sessions.map((item) => new SessionMentorBean(item));
                const sessionCollector = new SessionCollector<SessionMentorBean>(mentorSessions, from.toISOString(), 1);
                const isSameWeek = this.mdp.isSameWeek(from.toISOString(), this.props.sessionCollector.selectedDate);
                if (!this.state.sessionDetail.sessions || !!forceRefresh) {
                    const selectedDate = new Date(this.state.selectedDate);
                    this.refreshSession(selectedDate, sessionCollector);
                    newState = {
                        selectedDate: selectedDate.toISOString(),
                        sessionDetail: this.updateSessionDetail(selectedDate, sessionCollector)
                    }
                }
                if (isSameWeek || !forceRefresh) {
                    this.props.setSessionCollector(sessionCollector);
                    this.setState({
                        daysBar: this.updateDaysBar(counter),
                        loading: false,
                        ...newState
                    }, () => {
                        this.updateScrollTop();
                    })
                } else {
                    this.setState(newState, () => this.updateScrollTop())
                }
            }, () => {
                this.setState({
                    loading: false
                })
            })
    }

    private updateScrollTop() {
        const extraSpace = 150;
        const show = window.innerHeight + extraSpace < Utilities.getDocumentHeight();
        const sessionDetail = {
            ...this.state.sessionDetail,
            scrollTop: show
        };
        this.setState({
            sessionDetail
        })
    }

    private updateDaysBar(counter: number, sessionCollector?: SessionCollector<SessionMentorBean>) {
        if (!sessionCollector) {
            sessionCollector = this.props.sessionCollector;
        }
        return {
            counter,
            rangeDays: sessionCollector.getRangeDays(),
            weekDate: sessionCollector.selectedDate
        }
    }

    private updateSessionDetail(date: Date, sessionCollector?: SessionCollector<SessionMentorBean>) {
        const collector = sessionCollector || this.props.sessionCollector;
        return {
            ...this.state.sessionDetail,
            sessions: collector.getSessionsFrom(date.getDay())
        }
    }

    private refreshSession(date: Date, sessionCollector?: SessionCollector<SessionMentorBean>) {
        const collector = sessionCollector || this.props.sessionCollector;
        clearInterval(this.interval);
        this.interval = setInterval(() => {
            collector.updateCollector();
            this.setState({
                sessionDetail: {
                    ...this.state.sessionDetail,
                    sessions: collector.getSessionsFrom(date.getDay())
                }
            });
        }, 1100);
    }
}

const MentorHome: React.FC<{}> = () => {
    const firstMonday = Utilities.getMonday().toISOString();
    const [sessionCollector, setSessionCollector] =  React.useState(
        new SessionCollector<SessionMentorBean>([], firstMonday, 1));

    const noAttendedSessions = useHandlerNoAttendedSessions();

    return (
        <MentorHomeCore
            sessionCollector={sessionCollector}
            setSessionCollector={setSessionCollector}
            noAttendedSessions={noAttendedSessions} />
    )
}

export default MentorHome;
