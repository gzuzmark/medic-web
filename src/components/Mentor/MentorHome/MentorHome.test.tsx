import { shallow } from 'enzyme';
import 'jest-localstorage-mock';
import * as React from 'react';
import Utilities from "../../../common/Utils/Utilities";
import {SessionCollector} from "../../../domain/Session/SessionCollector";
import {SessionMentorBean} from "../../../domain/Session/SessionMentorBean";
import {MentorHomeCore} from './MentorHome';


describe('MentorHome Test',() => {
    let props: any;
    let mountedMentorHome: any;
    const firstMonday = Utilities.getMonday().toISOString();
    const getComponent = () => {
        if (!mountedMentorHome) {
            mountedMentorHome = shallow(
                <MentorHomeCore {...props} />
            );
        }
        return mountedMentorHome;
    };

    beforeEach(() => {
        const sessionCollector = new SessionCollector<SessionMentorBean>([], firstMonday, 1);
        props = {
            noAttendedSessions: {
                doRequest: () => void(0),
                item: null,
                loading: false
            },
            sessionCollector,
            setSessionCollector: (s: SessionCollector<SessionMentorBean>) => void(0)
        };
        mountedMentorHome = undefined;
    });

    it("event: loadSessions should have been called when getNewSessions is called", () => {
        const component = getComponent();
        const date = new Date();
        const spy = jest.spyOn(component.instance(), 'loadSessions');
        component.instance().getNewSessions(date.toISOString(), -1);
        expect(spy).toHaveBeenCalled();
    });
});
