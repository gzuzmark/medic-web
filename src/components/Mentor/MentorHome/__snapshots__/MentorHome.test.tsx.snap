// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MentorHome Test render: MentorHome render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MentorHome />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div>
          <div>
            <Icon
              name="calendar"
            />
            <styled.h2>
              Tus sesiones
            </styled.h2>
          </div>
        </div>,
        <DayHandlerBar
          loading={false}
          onChangeDate={[Function]}
          onChangeWeek={[Function]}
          selectedDate={undefined}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <Icon
              name="calendar"
            />
            <styled.h2>
              Tus sesiones
            </styled.h2>
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Icon
                name="calendar"
              />,
              <styled.h2>
                Tus sesiones
              </styled.h2>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "name": "calendar",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Tus sesiones",
              },
              "ref": null,
              "rendered": "Tus sesiones",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "loading": false,
          "onChangeDate": [Function],
          "onChangeWeek": [Function],
          "selectedDate": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div>
            <div>
              <Icon
                name="calendar"
              />
              <styled.h2>
                Tus sesiones
              </styled.h2>
            </div>
          </div>,
          <DayHandlerBar
            loading={false}
            onChangeDate={[Function]}
            onChangeWeek={[Function]}
            selectedDate={undefined}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <Icon
                name="calendar"
              />
              <styled.h2>
                Tus sesiones
              </styled.h2>
            </div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Icon
                  name="calendar"
                />,
                <styled.h2>
                  Tus sesiones
                </styled.h2>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "name": "calendar",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Tus sesiones",
                },
                "ref": null,
                "rendered": "Tus sesiones",
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "loading": false,
            "onChangeDate": [Function],
            "onChangeWeek": [Function],
            "selectedDate": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
